superset-frontend/src/explore/components/DataTablesPane/useGetResultsOrSamples.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { useState, useEffect } from 'react';
+import { t, ensureIsArray } from '@superset-ui/core';
+import { getChartDataRequest } from 'src/components/Chart/chartAction';
+import { getClientErrorObject } from 'src/utils/getClientErrorObject';
+import { requestPayload, responseData } from './types';
+
+const cache = new WeakMap();
+
+export function useGetResultsOrSamples({
+  isRequest,
+  resultType,
+  queryFormData,
+  datasource,
+  queryForce,
+  ownState,
+  actions,
+}: requestPayload): responseData {
+  // todo: this hook should split into useGetResults and useGetSamples after change endpoint
+  const [response, setResponse] = useState<responseData>({
```

Comment:
I think `useReducer` is a better choice for handling complex objects in a state. Alternatively, we could split that state into 5 `useState`s

Reply:
done, I have added a new `samples` endpoint and refactored this logic.